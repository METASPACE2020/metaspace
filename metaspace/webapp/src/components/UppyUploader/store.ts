// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from 'vue'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State | any>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach((listener) => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
