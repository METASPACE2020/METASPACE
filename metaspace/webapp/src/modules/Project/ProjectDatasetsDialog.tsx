import { defineComponent, computed, reactive } from '@vue/composition-api'
import { importDatasetsIntoProjectMutation, ProjectsListProject } from '../../api/project'
import { Dialog, Checkbox, Button } from '../../lib/element-ui'
import { uniqBy } from 'lodash'
import './ProjectDatasetsDialog.scss'
import { DatasetListItem, datasetListItemsQuery } from '../../api/dataset'
import { useQuery } from '@vue/apollo-composable'
import ElapsedTime from '../../components/ElapsedTime'
import reportError from '../../lib/reportError'

interface CheckOptions {
  [key: string]: boolean
}

interface ProjectDatasetsDialogState {
  selectedDatasets: CheckOptions
  isSubmitting: boolean
  hasChanged: boolean
}

interface ProjectDatasetsDialogProps {
  project: ProjectsListProject
  currentUserId: string
  visible: boolean
  isAdmin: boolean
  projectDatasets: string[]
  dialogLabel: string
  saveBtnLabel: string
  cancelBtnLabel: string
  selectAllLabel: string
  selectNoneLabel: string
  noDatasetsLabel: string
  refreshData: () => Promise<any>
}

export const ProjectDatasetsDialog = defineComponent<ProjectDatasetsDialogProps>({
  name: 'ProjectDatasetsDialog',
  props: {
    project: { type: Object, default: undefined },
    currentUserId: { type: String, default: undefined },
    dialogLabel: { type: String, default: 'Would you like to include/remove previously submitted datasets?' },
    saveBtnLabel: { type: String, default: 'Update' },
    cancelBtnLabel: { type: String, default: 'Cancel' },
    selectAllLabel: { type: String, default: 'Select all' },
    selectNoneLabel: { type: String, default: 'Select none' },
    noDatasetsLabel: { type: String, default: 'No datasets available' },
    visible: { type: Boolean, default: true },
    isAdmin: { type: Boolean, default: false },
    projectDatasets: { type: Array, default: () => [] },
    refreshData: { type: Function, required: true },
  },
  setup(props, ctx) {
    const { emit, root } = ctx
    const { $apollo } = root
    const getProjectDatasets = () => {
      const defaultDatasets = {} as CheckOptions
      if (Array.isArray(props.projectDatasets)) {
        props.projectDatasets.forEach((ds: any) => {
          defaultDatasets[ds.id] = true
        })
      }
      return defaultDatasets
    }
    const state = reactive<ProjectDatasetsDialogState>({
      selectedDatasets: getProjectDatasets(),
      isSubmitting: false,
      hasChanged: false,
    })
    const {
      result: datasetsResult,
      loading,
    } = useQuery<{allDatasets: DatasetListItem[]}>(datasetListItemsQuery, {
      dFilter: {
        submitter: props.currentUserId,
      },
    })
    const datasets = computed(() => datasetsResult.value != null
      ? datasetsResult.value.allDatasets as DatasetListItem[] : null)

    const getAvailableDatasets = (datasets: any) => {
      if (!props.isAdmin) {
        return datasets
      }

      if (Array.isArray(datasets) && Array.isArray(props.projectDatasets)) {
        // show not owned datasets if admin, so it can be removed
        return uniqBy(datasets.concat(props.projectDatasets), 'id')
      }

      return []
    }

    const handleClose = () => {
      emit('close')
    }

    const handleDatasetCheck = (value: boolean, key: string) => {
      state.selectedDatasets[key] = value
      if (!state.hasChanged) {
        state.hasChanged = true
      }
    }

    const handleSelectNone = () => {
      const availableDatasets = getAvailableDatasets(datasets?.value).map((ds: any) => ds.id)
      availableDatasets.forEach((key: string) => { state.selectedDatasets[key] = false })
      if (!state.hasChanged) {
        state.hasChanged = true
      }
    }

    const handleSelectAll = () => {
      const availableDatasets = getAvailableDatasets(datasets?.value).map((ds: any) => ds.id)
      availableDatasets.forEach((key: string) => { state.selectedDatasets[key] = true })
      if (!state.hasChanged) {
        state.hasChanged = true
      }
    }

    const handleProjectDatasetsUpdate = async() => {
      const previousDatasets = getProjectDatasets()
      const removedDatasetIds: string[] = []
      const addedDatasetIds: string[] = []
      state.isSubmitting = true
      Object.keys(state.selectedDatasets).forEach((key: string) => {
        if (previousDatasets[key] && !state.selectedDatasets[key]) { // remove ds
          removedDatasetIds.push(key)
        } else if (!previousDatasets[key] && state.selectedDatasets[key]) { // new ds * do not add already
          addedDatasetIds.push(key)
        }
      })

      try {
        if (addedDatasetIds.length > 0 || removedDatasetIds.length > 0) {
          await $apollo.mutate({
            mutation: importDatasetsIntoProjectMutation,
            variables: { projectId: props.project?.id, datasetIds: addedDatasetIds, removedDatasetIds },
          })
        }
        await props.refreshData()
        emit('update')
      } catch (err) {
        reportError(err)
      } finally {
        state.isSubmitting = false
      }
    }

    return () => {
      const {
        visible, project, dialogLabel, saveBtnLabel, cancelBtnLabel,
        selectAllLabel, selectNoneLabel, noDatasetsLabel,
      } = props
      const { name } = project
      const availableDatasets = getAvailableDatasets(datasets?.value)

      return (
        <Dialog
          class='project-datasets-dialog'
          visible={visible}
          append-to-body
          title={`Manage ${name}\`s datasets`}
          lockScroll={false}
          onClose={handleClose}>
          <div class="mt-6">
            <h4 class="m-0">
              {dialogLabel}
            </h4>
            {
              Array.isArray(availableDatasets) && availableDatasets.length > 0
                && <div class="dataset-checkbox-list leading-6">
                  <div class="mb-2">
                    <span
                      class="select-link"
                      onClick={handleSelectNone}>{selectNoneLabel}</span>
                    <span> | </span>
                    <span
                      class="select-link"
                      onClick={handleSelectAll}>{selectAllLabel}</span>
                  </div>
                  {
                    availableDatasets.map((dataset) => <Checkbox
                      class="flex h-6 items-center m-0 mx-2"
                      key={dataset.id}
                      value={state?.selectedDatasets[dataset.id]}
                      disabled={state?.isSubmitting}
                      onChange={(value: boolean) => handleDatasetCheck(value, dataset.id)}
                      label={dataset.name}>
                      <span class="truncate">
                        { dataset?.name }
                      </span>
                      <span class="text-gray-700 text-xs tracking-wide pl-1">
                        <ElapsedTime date={dataset?.uploadDT} />
                      </span>
                    </Checkbox>)
                  }
                </div>
            }
            {
              Array.isArray(availableDatasets) && availableDatasets.length === 0
                && <div class="flex items-center justify-center leading-6">
                  <div>{noDatasetsLabel}</div>
                </div>
            }

            <div class="button-bar">
              <Button onClick={handleClose}>
                {cancelBtnLabel}
              </Button>
              <Button
                class='w-32'
                loading={state?.isSubmitting}
                disabled={!state?.hasChanged}
                type="primary"
                onClick={handleProjectDatasetsUpdate}>
                {saveBtnLabel}
              </Button>
            </div>
          </div>
        </Dialog>
      )
    }
  },
})
