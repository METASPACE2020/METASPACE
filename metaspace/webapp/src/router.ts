/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter, { RawLocation } from 'vue-router'
import AboutPage from './modules/App/AboutPage.vue'
import DatasetsPage from './modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from './modules/Account'
import { redirectAfterSignIn } from './modules/Account/signInReturnUrl'
import { updateDBParam } from './modules/Filters/url'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Annotations/AnnotationsPage.vue'),
  DatasetSummary: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ './modules/Datasets/summary/DatasetSummary.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ './modules/MetadataEditor/MetadataEditPage.vue'),
  ImageAlignmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ './modules/ImageAlignment/ImageAlignmentPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ './modules/MetadataEditor/UploadPage.vue'),
  DesignPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DesignPage" */ './design/DesignPage.vue'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/HelpPage.vue'),
  EditUserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/GroupProfile/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/GroupsListPage.vue'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/TermsPage.vue'),
  PublicationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PublicationsPage.vue'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/GroupProfile/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Project/ViewProjectPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(about)?', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
    { path: '/annotations', component: asyncPages.AnnotationsPage },
    {
      path: '/datasets',
      component: DatasetsPage,
      children: [
        { path: '', component: asyncPages.DatasetTable },
        { path: 'summary', component: asyncPages.DatasetSummary },
      ],
    },
    { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
    { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.ImageAlignmentPage },
    { path: '/upload', component: asyncPages.UploadPage },
    { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },
    { path: '/user/me', component: asyncPages.EditUserPage },

    { path: '/admin/health', component: asyncPages.SystemHealthPage },
    { path: '/admin/groups', component: asyncPages.GroupsListPage },

    { path: '/account/sign-in', component: DialogPage, props: { dialog: 'signIn' } },
    { path: '/account/sign-in-success', redirect: redirectAfterSignIn },
    { path: '/account/create-account', component: DialogPage, props: { dialog: 'createAccount' } },
    { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
    { path: '/account/reset-password', component: ResetPasswordPage },

    { path: '/group/create', component: asyncPages.CreateGroupPage },
    { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

    { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
    { // Legacy URL sent in "request access" emails up until Feb 2019
      path: '/project/:projectIdOrSlug/manage',
      redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } } as RawLocation,
    },
    { path: '/projects', component: asyncPages.ProjectsListPage },

    { path: '/terms', component: asyncPages.TermsPage, meta: { footer: true } },
    { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
    { path: '/publications', component: asyncPages.PublicationsPage, meta: { footer: true } },

    { path: '/design', component: asyncPages.DesignPage, meta: { footer: true, flex: true } },

    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
