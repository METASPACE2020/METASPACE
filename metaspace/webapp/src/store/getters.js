import {mzFilterPrecision} from '../util';
import {decodeParams, decodeSettings, getLevel} from '../modules/Filters';
import config from '../config';

/** For filters where empty string is a valid client-side value that must be converted to empty string for the API */
const noneToEmptyString = s => s === 'none' ? '' : s;

export default {
  filterLevel(state) {
    return getLevel(state.route.path);
  },

  filter(state) {
    return decodeParams(state.route, state.filterLists);
  },

  settings(state) {
    return decodeSettings(state.route);
  },

  ftsQuery(state, getters) {
    return getters.filter.simpleQuery;
  },

  gqlAnnotationFilter(state, getters) {
    const filter = getters.filter;
    const colocalizationAlgo = getters.settings.annotationView.colocalizationAlgo;

    const f = {
      database: filter.database,
      compoundQuery: filter.compoundName,
      chemMod: noneToEmptyString(filter.chemMod),
      neutralLoss: noneToEmptyString(filter.neutralLoss),
      adduct: filter.adduct,
      fdrLevel: filter.fdrLevel,
      colocalizedWith: filter.colocalizedWith,
      // Only include colocalizationAlgo if there is another filter that uses it. Otherwise the annotations list
      // refreshes unnecessarily when changing algorithm.
      colocalizationAlgo: filter.colocalizedWith || filter.colocalizationSamples ? colocalizationAlgo : null,
      colocalizationSamples: filter.colocalizationSamples,
      offSample: filter.offSample == null ? undefined : !!filter.offSample,
    };

    if (!config.features.neutral_losses && !config.features.advanced_ds_config) {
      f.hasNeutralLoss = false;
    }

    if (!config.features.chem_mods && !config.features.advanced_ds_config) {
      f.hasChemMod = false;
    }

    if (!config.features.all_adducts) {
      f.hasHiddenAdduct = false;
    }

    if (filter.minMSM)
      f.msmScoreFilter = {min: filter.minMSM, max: 1.0};

    if (filter.mz) {
      const mz = parseFloat(filter.mz),
            deltamz = parseFloat(mzFilterPrecision(mz));
      f.mzFilter = {
        min: mz - deltamz,
        max: mz + deltamz
      };
    }

    return f;
  },

  gqlDatasetFilter(state, getters) {
    const filter = getters.filter;
    const {group, project, submitter, datasetIds, polarity,
           organism, organismPart, condition, growthConditions,
           ionisationSource, analyzerType, maldiMatrix, metadataType,
           compoundName} = filter;
    const level = getters.filterLevel;
    const hasAnnotationMatching = level === 'dataset' && compoundName ? { compoundQuery: compoundName } : undefined;
    return {
      group: group,
      project: project,
      submitter: submitter,

      // temporary workaround because of array-related bugs in apollo-client
      ids: datasetIds ? datasetIds.join("|") : null,

      organism,
      organismPart,
      condition,
      growthConditions,
      ionisationSource,
      maldiMatrix,
      analyzerType,
      polarity: polarity ? polarity.toUpperCase() : null,
      metadataType,
      hasAnnotationMatching,
    }
  },

  gqlColocalizationFilter(state, getters) {
    const {datasetIds, colocalizedWith, database, fdrLevel} = getters.filter;
    const colocalizationAlgo = getters.settings.annotationView.colocalizationAlgo;
    if (datasetIds && !datasetIds.includes('|') && colocalizedWith != null && database != null && fdrLevel != null) {
      return {colocalizedWith, colocalizationAlgo, database, fdrLevel};
    } else {
      return null;
    }
  },
}
