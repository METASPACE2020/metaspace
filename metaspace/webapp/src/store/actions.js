import { fetchOptionListsQuery } from '../api/metadata'
import { decodeParams } from '../modules/Filters'
import { computed } from 'vue'
import store from '../store/index'

export default {
  async initFilterLists(context, apolloClient) {
    if (context.state.filterListsLoading || context.state.filterLists != null) return
    await context.commit('setFilterListsLoading')

    const response = await apolloClient.query({
      query: fetchOptionListsQuery,
      fetchPolicy: 'cache-first',
    })

    // set annotationIds default according to values passed from snapshot to the store state
    context.commit('setFilterLists', {
      ...response.data,
      annotationIds: computed(() => store.state.snapshotAnnotationIds),
    })

    // Refresh the current filter so that computed defaults that depend on `filterLists` are applied
    const filter = decodeParams(context.state.route, context.state.filterLists)

    context.commit('updateFilter', filter)
  },
}
