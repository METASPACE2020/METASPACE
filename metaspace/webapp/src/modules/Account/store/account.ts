import { Module } from 'vuex';
import { DialogType } from '../dialogs';
import router from '../../../router';

export interface AccountState {
  dialog: DialogType | null;
  showDialogAsPage: boolean;
}

const account = {
  namespaced: true,
  state: {
    dialog: null,
    showDialogAsPage: false
  },
  mutations: {
    showDialog(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = false;
    },
    showDialogAsPage(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = true;
    },
    hideDialog(state: AccountState, dialog: DialogType) {
      if (state.dialog === dialog) {
        state.dialog = null;
        if (state.showDialogAsPage) {
          router.push('/');
        }
      }
    }
  }
};

export default account as Module<AccountState, any>;
