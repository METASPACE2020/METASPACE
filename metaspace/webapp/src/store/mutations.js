import {isEqual, pull, without} from 'lodash-es';
import router from '../router';
import compare from '../lib/compare';

import {
  decodeParams,
  encodeParams, encodeSections, encodeSortOrder, FILTER_SPECIFICATIONS,
  stripFilteringParams,
} from '../modules/Filters';
import {getFilterInitialValue} from '../modules/Filters';


function updatedLocation(state, filter) {
  let query = encodeParams(filter, state.route.path, state.filterLists);

  state.lastUsedFilters[state.route.path] = {
    filter,
    query,
    order: state.orderedActiveFilters
  };

  return {
    query: Object.assign(query, stripFilteringParams(state.route.query))
  };
}

function replaceURL(state, filter) {
  router.replace(updatedLocation(state, filter));
}

function pushURL(state, filter) {
  router.push(updatedLocation(state, filter));
}

function sortFilterKeys(keys) {
  keys.sort((a, b) => compare(FILTER_SPECIFICATIONS[a].sortOrder || 100, FILTER_SPECIFICATIONS[b].sortOrder || 100));
}

export default {
  updateFilter (state, filter) {
    let active = [];

    // drop unset filters
    for (var i = 0; i < state.orderedActiveFilters.length; i++) {
      let key = state.orderedActiveFilters[i];
      if (filter[key] !== undefined)
        active.push(key);
    }

    // append newly added filters to the end
    for (var key in filter)
      if (filter[key] !== undefined &&
          active.indexOf(key) == -1)
        active.push(key);

    // sort
    sortFilterKeys(active);

    const changedFilterSet = !isEqual(state.orderedActiveFilters, active);

    state.orderedActiveFilters = active;
    if (changedFilterSet)
      pushURL(state, filter);
    else
      replaceURL(state, filter);
  },

  addFilter (state, name) {
    const initialValue = getFilterInitialValue(name, state.filterLists);

    // FIXME: is there any way to access getters here?
    let filter = Object.assign(decodeParams(state.route, state.filterLists),
                               {[name]: initialValue});

    state.orderedActiveFilters.push(name);
    pushURL(state, filter);
  },

  updateFilterOnNavigate(state, to) {
    const newActiveFilters = Object.keys(decodeParams(to, state.filterLists));

    const removed = without(state.orderedActiveFilters, ...newActiveFilters);
    const added = without(newActiveFilters, ...state.orderedActiveFilters);

    pull(state.orderedActiveFilters, ...removed);
    state.orderedActiveFilters.push(...added);
    sortFilterKeys(state.orderedActiveFilters);
  },

  setFilterListsLoading(state) {
    state.filterListsLoading = true;
  },

  setFilterLists(state, filterLists) {
    state.filterLists = filterLists;
    state.filterListsLoading = false;
  },

  setAnnotation(state, annotation) {
    state.annotation = annotation;
  },

  updateAnnotationTableStatus(state, isLoading) {
    state.tableIsLoading = isLoading;
  },

  startTour(state, tourDescription) {
    state.currentTour = tourDescription;
  },

  endTour(state) {
    state.currentTour = null;
  },

  updateAnnotationViewSections(state, activeSections) {
    let query = Object.assign({}, state.route.query, {
      sections: encodeSections(activeSections)
    });
    router.replace({query});
  },

  setColormap(state, colormap) {
    let query = Object.assign({}, state.route.query, {
      cmap: colormap
    });
    router.replace({query});
  },

  setCurrentPage(state, page) {
    let query = Object.assign({}, state.route.query, {page});
    router.replace({query});
  },

  setSortOrder(state, sortOrder) {
    let query = Object.assign({}, state.route.query, {
      sort: encodeSortOrder(sortOrder)
    });
    router.replace({query});
  },

  setDatasetTab(state, tab) {
    const path = {
      'list': '/datasets',
      'summary': '/datasets/summary'
    }[tab.toLowerCase()];
    router.replace({
      path,
      query: state.route.query,
    });
  }
};
