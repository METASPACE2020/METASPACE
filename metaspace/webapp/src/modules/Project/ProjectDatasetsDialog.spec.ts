import { mount } from '@vue/test-utils'
import router from '../../router'
import store from '../../store/index'
import Vue from 'vue'
import Vuex from 'vuex'
import { sync } from 'vuex-router-sync'
import { ProjectDatasetsDialog } from './ProjectDatasetsDialog'
import { initMockGraphqlClient, apolloProvider } from '../../../tests/utils/mockGraphqlClient'

Vue.use(Vuex)
sync(store, router)

describe('ProjectDatasetsDialog', () => {
  const propsData = {
    project: {
      id: '05c6519a-8049-11eb-927e-6bf28a9b25ae',
      name: 'Test',
      currentUserRole: 'MANAGER',
    },
    currentUserId: '039801c8-919e-11eb-908e-3b2b8e672707',
    dialogLabel: 'Would you like to include/remove previously submitted datasets?',
    saveBtnLabel: 'Update',
    cancelBtnLabel: 'Cancel',
    selectAllLabel: 'Select all',
    selectNoneLabel: 'Select none',
    noDatasetsLabel: 'No datasets available',
    visible: true,
    isAdmin: true,
    refreshData: () => {},
    projectDatasets: [
      {
        id: '2021-03-31_11h02m28s',
        name: 'New 3 (1)',
        status: 'FINISHED',
        canDownload: true,
        uploadDT: '2021-03-31T14:02:28.722Z',
      },
      {
        id: '2021-03-30_18h25m18s',
        name: 'Untreated_3_434_super_lite_19_31 (1)',
        status: 'FINISHED',
        statusUpdateDT: '2021-03-30T21:29:02.043Z',
        uploadDT: '2021-03-30T21:25:18.473Z',
      },
    ],
  }
  const propsDataNoDsSelected = {
    project: {
      id: '05c6519a-8049-11eb-927e-6bf28a9b25ae',
      name: 'Test',
      currentUserRole: 'MANAGER',
    },
    currentUserId: '039801c8-919e-11eb-908e-3b2b8e672707',
    dialogLabel: 'Would you like to include/remove previously submitted datasets?',
    saveBtnLabel: 'Update',
    cancelBtnLabel: 'Cancel',
    selectAllLabel: 'Select all',
    selectNoneLabel: 'Select none',
    noDatasetsLabel: 'No datasets available',
    visible: true,
    isAdmin: true,
    refreshData: () => {},
    projectDatasets: [],
  }

  const testHarness = Vue.extend({
    components: {
      ProjectDatasetsDialog,
    },
    render(h) {
      return h(ProjectDatasetsDialog, { props: this.$attrs })
    },
  })
  const graphqlWithData = () => {
    initMockGraphqlClient({
      Query: () => ({
        allDatasets: () => {
          return [{
            id: '2021-03-31_11h02m28s',
            name: 'New 3 (1)',
            uploadDT: '2021-03-31T14:02:28.722Z',
            __typename: 'Dataset',
          }, {
            id: '2021-03-30_18h25m18s',
            name: 'Untreated_3_434_super_lite_19_31 (1)',
            uploadDT: '2021-03-30T21:25:18.473Z',
            __typename: 'Dataset',
          }]
        },
      }),
    })
  }
  const graphqlWithNoData = () => {
    initMockGraphqlClient({
      Query: () => ({
        allDatasets: () => {
          return []
        },
      }),
    })
  }

  it('it should match snapshot', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  it('it should match no dataset snapshot', async() => {
    graphqlWithNoData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  it('it should have disabled update button when no data available', async() => {
    graphqlWithNoData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)
  })

  it('it should have disabled update button when no dataset selection has changed', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)
  })

  it('it should uncheck all datasets when clicking select none', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    const selectNone = wrapper.findAll('.select-link').at(0)
    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)
    selectNone.trigger('click')
    await Vue.nextTick()

    wrapper.findAll('.el-checkbox').wrappers.forEach((checkBox) => {
      expect(checkBox.classes('is-checked')).toBe(false)
    })

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(false)
  })

  it('it should check all datasets when clicking select all', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    const selectAll = wrapper.findAll('.select-link').at(1)
    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)
    selectAll.trigger('click')
    await Vue.nextTick()

    wrapper.findAll('.el-checkbox').wrappers.forEach((checkBox) => {
      expect(checkBox.classes('is-checked')).toBe(true)
    })

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(false)
  })

  it('it should check all datasets when clicking select all', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    const selectAll = wrapper.findAll('.select-link').at(1)
    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)
    selectAll.trigger('click')
    await Vue.nextTick()

    wrapper.findAll('.el-checkbox').wrappers.forEach((checkBox) => {
      expect(checkBox.classes('is-checked')).toBe(true)
    })

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(false)
  })

  it('it should check one dataset and hit update', async() => {
    graphqlWithData()

    const wrapper = mount(testHarness, {
      store,
      router,
      propsData: {
        ...propsDataNoDsSelected,
      },
    })
    await Vue.nextTick()

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)

    wrapper.findAll('.el-checkbox').wrappers.forEach((checkBox) => {
      expect(checkBox.classes('is-checked')).toBe(false)
    })

    wrapper.findAll('.el-checkbox').at(0).trigger('click')
    await Vue.nextTick()

    expect(wrapper.findAll('.el-checkbox').at(0).classes('is-checked')).toBe(true)
    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(false)
  })

  it('it should uncheck one dataset and hit update', async() => {
    graphqlWithData()

    const wrapper = mount(testHarness, {
      store,
      router,
      propsData,
    })
    await Vue.nextTick()

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)

    wrapper.findAll('.el-checkbox').wrappers.forEach((checkBox) => {
      expect(checkBox.classes('is-checked')).toBe(true)
    })

    wrapper.findAll('.el-checkbox').at(0).trigger('click')
    await Vue.nextTick()

    expect(wrapper.findAll('.el-checkbox').at(0).classes('is-checked')).toBe(false)
    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(false)
  })
})
