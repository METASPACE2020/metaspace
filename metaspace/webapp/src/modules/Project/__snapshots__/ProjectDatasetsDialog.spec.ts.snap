// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProjectDatasetsDialog > it should match snapshot 1`] = `
"<div class=\\"mock-el-dialog project-datasets-dialog\\" modelvalue=\\"true\\" append-to-body=\\"true\\" title=\\"Manage Test\`s datasets\\" lockscroll=\\"false\\" props=\\"[object Object]\\">
  <div class=\\"mt-6\\">
    <h4 class=\\"m-0\\">Would you like to include/remove previously submitted datasets?</h4>
    <div class=\\"dataset-checkbox-list leading-6\\">
      <div class=\\"mb-2\\"><span class=\\"select-link\\">Select none</span><span> | </span><span class=\\"select-link\\">Select all</span></div><label class=\\"el-checkbox flex h-6 items-center m-0 mx-2\\"><span class=\\"el-checkbox__input\\"><input class=\\"el-checkbox__original\\" type=\\"checkbox\\" value=\\"New 3 (1)\\"><span class=\\"el-checkbox__inner\\"></span></span><span class=\\"el-checkbox__label\\"><span class=\\"truncate\\">New 3 (1)</span><span class=\\"text-gray-700 text-xs tracking-wide pl-1\\"><span class=\\"sm-elapsed-time\\" title=\\"2021/03/31, 14:02\\">a year ago</span></span>
        <!--v-if--></span>
      </label><label class=\\"el-checkbox flex h-6 items-center m-0 mx-2\\"><span class=\\"el-checkbox__input\\"><input class=\\"el-checkbox__original\\" type=\\"checkbox\\" value=\\"Untreated_3_434_super_lite_19_31 (1)\\"><span class=\\"el-checkbox__inner\\"></span></span><span class=\\"el-checkbox__label\\"><span class=\\"truncate\\">Untreated_3_434_super_lite_19_31 (1)</span><span class=\\"text-gray-700 text-xs tracking-wide pl-1\\"><span class=\\"sm-elapsed-time\\" title=\\"2021/03/30, 21:25\\">a year ago</span></span>
        <!--v-if--></span>
      </label>
    </div>
    <!---->
    <div class=\\"button-bar\\"><button ariadisabled=\\"false\\" type=\\"button\\" class=\\"el-button\\">
        <!--v-if--><span class=\\"\\">Cancel</span>
      </button><button ariadisabled=\\"true\\" disabled=\\"\\" type=\\"button\\" class=\\"el-button el-button--primary is-disabled w-32\\">
        <!--v-if--><span class=\\"\\">Update</span>
      </button></div>
  </div>
</div>"
`;
